### ISSUE TITLE
RFC: shifted reduction

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#119929]
### 🚀 The feature, motivation and pitch

# Context
Consider reduction on the last dimension of a 2D contiguous tensor with shape [M, N]. If N has bad alignment (e.g. 2 ** 20 - 1), then most rows of the tensor will start at an unaligned address and makes the memory accesses inefficient in GPU.

Here is an example to show this with some numbers:
```
def run_test(x, r, op):
    def f(arg):
        return getattr(arg, op)(dim=-1)

    arg = torch.randn(x, r)
    opt_f = torch.compile(f)
    ms = do_bench(lambda: opt_f(arg))
    print(f"latency {ms} ms")
    act = opt_f(arg)
    ref = f(arg)

    tol = {"atol": 1e-4, "rtol": 1e-4}
    assert torch.allclose(ref, act, **tol), f"ref\n{ref}\nact\n{act}"
    print("bye")


run_test(1024, 2 ** 20, op="sum") # aligned shape results in 2.69ms
run_test(1024, 2 ** 20 - 1, op="sum") # unaligned shape results in 3.38ms
```

From above we see that processing a slightly smaller tensor input with unaligned shape is 3.38ms / 2.69ms = 1.26x slower than a slightly larger tensor with aligned shape!

# Shifted Reduction

The inductor generated triton code for the unaligned reduction looks like:
```
@triton.jit
def triton_(in_ptr0, out_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 1024
    rnumel = 1048575
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    _tmp2 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (1048575*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
        tmp3 = _tmp2 + tmp1
        _tmp2 = tl.where(rmask & xmask, tmp3, _tmp2)
    tmp2 = tl.sum(_tmp2, 1)[:, None]
    tl.store(out_ptr0 + (x0), tmp2, xmask)
```

Check the tl.load call. The address offset is `(r1 + (1048575 * x0))` . This value will be  a multiple of ALIGN (say 16), only if x0 is a multiple of ALIGN which only happens `1/ALIGN` percent of time. But one thing we can do is to shift the address for the load a bit and change the code to (especially check the 3 lines of code change with comments):
```
@triton.jit
def triton_(in_ptr0, out_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 1024
    rnumel = 1048575
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    _tmp2 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    tmpv = -((xoffset * 1048575) % 16) # compute a negative offset
    for roffset in range(tmpv, rnumel, RBLOCK): # roffset may start from a negative value to make the computed address for load aligned
        rindex = roffset + rbase
        rmask = rindex < rnumel and rindex >= 0 # need skip negative indices
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (1048575*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
        tmp3 = _tmp2 + tmp1
        _tmp2 = tl.where(rmask & xmask, tmp3, _tmp2)
    tmp2 = tl.sum(_tmp2, 1)[:, None]
    tl.store(out_ptr0 + (x0), tmp2, xmask)
```

This make sure each load use an aligned address. With this change we improve the latency from 3.38ms to 2.87ms.

# Next steps
The above is the rough idea (thanks Jason and Horace for the valuable discussion), to make it more solid we need handle the following things

1. the change assumes that XBLOCK is 1. This is true for most inner reduction but we need handle it properly in inductor codegen
2. With shifted reduction, the improved perf is still a bit slower compared to using an aligned shape. Need dig more
3. For some shapes, shifted reduction can be slower than out of the box.

E.g.
run_test(1024, 2 ** 13, op = "softmax") # takes 58.5us
run_test(1024, 2 ** 13 - 1, op = "softmax") # takes 73us OOB and takes 80.3us with shifted reduction.

This probably is because we may run the reduction loop one more time compared to out of the box behavior. But I'll also need dig more.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @jansel @Chillee @eellison

**Comment by shunting314** [comment]
Comprehensive padding is implemented which is better than this prototype.
